### PR TITLE
Ignore lines between anon inner classes

### PR DESCRIPTION
**Pull request by JimDeanSpivey**
Issue: #653

2 caveats with this code I want to address

1. This implementation will have a slight, slight bug with double nested anonymous inner types. I just don't feel that writing all the code to make this check recursively check nested anonymous inner types is worth it, and it would make the code maintenance more of a burden to support a use-case that shouldn't exist (supporting double nested anonymous inner classes, really? :trollface: ).
2. ~~`EmptyLineCheck` is now no longer thread safe as it has to hold state of the lines which contain anonymous inner classes and method bodies. This is just a result of the fact that the API into checkstyle's is exposed as a streaming parsing API, as opposed to a fully loaded DOM. If it were DOM based, I could just setup the immutable objects ahead of time and let it all be sorted in a cleaner way. But it's not, so this is a caveat. The checkstyle API should certainly **not** be reusing plugin instances of the `Check` class, it should be created a `new` object per each file it validates with the `EmptyLineCheck` plugin. Provided that they are *not* reusing instances, no problems will exist concerning parallel access. As every `EmptyLinesCheck` will be unique and there is no shared access concern at all.~~ It is thread safe now due to the usage of thread locals.

Lastly, regarding caveat 2, I built the project in parallel if that helps, it passed:
`mvn clean install -Pqulice -T 4`

### CODE REVIEW

**Comment by JimDeanSpivey** [comment]
@davvd Okay, this time it is ready for a PR. I enhanced the unit tests to catch all the edge-cases that the original code wasn't catching for so I'm sure this one is ready.

**Comment by davvd** [comment]
@jimdeanspivey Thanks, someone will review your pull request soon

**Comment by davvd** [comment]
@pinaf this pull request is for you, please review

File: qulice-checkstyle/src/main/java/com/qulice/checkstyle/EmptyLinesCheck.java [hunk 1]
```diff
@@ -42,8 +42,13 @@
  * your code and make it more cohesive and readable. The bottom line is
  * that every method should look solid and do just <b>one thing</b>.
  *
+ * This class is thread safe. It relies on building a list of line ranges by
```
Comment (pinaf) [inline qulice-checkstyle/src/main/java/com/qulice/checkstyle/EmptyLinesCheck.java]: @JimDeanSpivey you said it is not thread safe in the PR. Which one is it?

File: qulice-checkstyle/src/main/java/com/qulice/checkstyle/LineRange.java [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+     * @param first The alpha line number.
+     * @param last The omega line number.
+     */
+    public LineRange(final int first, final int last) {
+        this.alpha = first;
+        this.omega = last;
+    }
+
+    /**
+     * Is the proposed line number within range.
```
Comment (pinaf) [inline qulice-checkstyle/src/main/java/com/qulice/checkstyle/LineRange.java]: @JimDeanSpivey let's change `proposed` to `given` here and in the the next 2 lines.

File: qulice-checkstyle/src/main/java/com/qulice/checkstyle/LineRange.java [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+     * @param line The proposed line number to check.
+     * @return True if the proposed line number is within range.
+     */
+    public boolean within(final int line) {
+        return line >= this.first() && line <= this.last();
+    }
+
+    /**
+     * Is the proposed range entirely within the LineRange.
```
Comment (pinaf) [inline qulice-checkstyle/src/main/java/com/qulice/checkstyle/LineRange.java]: @JimDeanSpivey same here. `proposed` -> `given`. also, entirely within `this range`.

File: qulice-checkstyle/src/main/java/com/qulice/checkstyle/LineRange.java [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+     * Example, tiven a LineRange of [10, 50], the proposed range of
+     * [12,48] should be within side that. And the method should return true.
+     * @param range The proposed LineRange to check.
+     * @return True if the proposed is entirely within LineRange.
+     */
+    public boolean within(final LineRange range) {
+        return range != null
```
Comment (pinaf) [inline qulice-checkstyle/src/main/java/com/qulice/checkstyle/LineRange.java]: @JimDeanSpivey no need for this null check. we should never be using nulls, ever :)

File: qulice-checkstyle/src/main/java/com/qulice/checkstyle/LineRange.java [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+            && range.first() >= this.first()
+            && range.last()  <= this.last();
```
Comment (pinaf) [inline qulice-checkstyle/src/main/java/com/qulice/checkstyle/LineRange.java]: @JimDeanSpivey extra whitespace here that we can remove

File: qulice-checkstyle/src/main/java/com/qulice/checkstyle/LineRanges.java [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+    /**
+     * Adds a line range to the internal collection.
+     * @param line The line range to add to the collection
+     */
+    public void add(final LineRange line) {
+        this.lines.collection().add(line);
+    }
+
+    /**
+     * Expses the iterator of the internal collection.
```
Comment (pinaf) [inline qulice-checkstyle/src/main/java/com/qulice/checkstyle/LineRanges.java]: @JimDeanSpivey let's not talk about the `internal collection` since that is an implementation detail. instead let's maintain encapsulation even in the javadocs :)
How about `Returns an iterator for this set.`?

File: qulice-checkstyle/src/main/java/com/qulice/checkstyle/LineRanges.java [hunk 7]
```diff
@@ -0,0 +1,153 @@
...
+     * @return Iterator pointing to the internal collections elements.
+     */
+    public Iterator<LineRange> iterate() {
```
Comment (pinaf) [inline qulice-checkstyle/src/main/java/com/qulice/checkstyle/LineRanges.java]: @JimDeanSpivey let's rename to `iterator`

File: qulice-checkstyle/src/main/java/com/qulice/checkstyle/LineRanges.java [hunk 8]
```diff
@@ -0,0 +1,153 @@
...
+public final class LineRanges {
+
+    /**
+     * ArrayList of line ranges.
+     */
+    private final transient LineRanges.LocalCollection lines =
+        new LineRanges.LocalCollection();
+
+    /**
+     * Adds a line range to the internal collection.
```
Comment (pinaf) [inline qulice-checkstyle/src/main/java/com/qulice/checkstyle/LineRanges.java]: @JimDeanSpivey same here, no mention of `internal collection`. `Adds a line range to this set` is fine.

File: qulice-checkstyle/src/main/java/com/qulice/checkstyle/LineRanges.java [hunk 9]
```diff
@@ -0,0 +1,153 @@
...
+    /**
+     * Expses the iterator of the internal collection.
+     * @return Iterator pointing to the internal collections elements.
+     */
+    public Iterator<LineRange> iterate() {
+        return this.lines.collection().iterator();
+    }
+
+    /**
+     * Detects if the proposed line number is with any of the line ranges.
```
Comment (pinaf) [inline qulice-checkstyle/src/main/java/com/qulice/checkstyle/LineRanges.java]: @JimDeanSpivey typo `with any`.

Comment (pinaf) [inline qulice-checkstyle/src/main/java/com/qulice/checkstyle/LineRanges.java]: @JimDeanSpivey same thing about `proposed`. let's go with `given`.

File: qulice-checkstyle/src/main/java/com/qulice/checkstyle/LineRanges.java [hunk 10]
```diff
@@ -0,0 +1,153 @@
...
+     * @return True if the proposed line number is within any line range.
+     */
+    public boolean inRange(final int line) {
+        return !this.lines.collection().isEmpty()
+            && FluentIterable.from(this.lines.collection())
+            .anyMatch(new LineRanges.LineWithAny(line));
+    }
+
+    /**
+     * Gets the subset of LineRanges that are within all proposed ranges. Does
```
Comment (pinaf) [inline qulice-checkstyle/src/main/java/com/qulice/checkstyle/LineRanges.java]: @JimDeanSpivey same thing about proposed. let's go with given.

File: qulice-checkstyle/src/main/java/com/qulice/checkstyle/LineRanges.java [hunk 11]
```diff
@@ -0,0 +1,153 @@
...
+     * not return null; instead, returns empty range if there are no matches.
+     * @param ranges The ranges to filter on.
+     * @return Returns all LineRange elements that are within range.
+     */
+    public LineRanges within(final LineRanges ranges) {
+        final LineRanges result = new LineRanges();
+        final Iterator<LineRange> iterator = ranges.iterate();
+        while (iterator.hasNext()) {
```
Comment (pinaf) [inline qulice-checkstyle/src/main/java/com/qulice/checkstyle/LineRanges.java]: @JimDeanSpivey can we go with a foreach? `for (final LineRange range : ranges) {`?

File: qulice-checkstyle/src/main/java/com/qulice/checkstyle/LineRanges.java [hunk 12]
```diff
@@ -0,0 +1,153 @@
...
+                if (next.within(line)) {
+                    result.add(line);
+                }
+            }
+        }
+        return result;
+    }
+
+    /**
+     * Clears the internal collection.
```
Comment (pinaf) [inline qulice-checkstyle/src/main/java/com/qulice/checkstyle/LineRanges.java]: @JimDeanSpivey `Clears this set`.

File: qulice-checkstyle/src/main/java/com/qulice/checkstyle/LineRanges.java [hunk 13]
```diff
@@ -0,0 +1,153 @@
...
+         * Default constructor.
+         * @param line The proposed line to check against all the line ranges.
+         */
+        private LineWithAny(final int line) {
+            this.proposed = line;
+        }
+
+        @Override
+        public boolean apply(final LineRange range) {
+            return range != null && range.within(this.proposed);
```
Comment (pinaf) [inline qulice-checkstyle/src/main/java/com/qulice/checkstyle/LineRanges.java]: @JimDeanSpivey no need for null checks

File: qulice-checkstyle/src/test/java/com/qulice/checkstyle/CheckstyleValidatorTest.java [hunk 14]
```diff
@@ -470,6 +470,49 @@ public void doesNotRejectUrlsInLongLines() throws Exception {
     }
 
     /**
+     * Ensures that spaces can exist outside of of methods that are within
```
Comment (pinaf) [inline qulice-checkstyle/src/test/java/com/qulice/checkstyle/CheckstyleValidatorTest.java]: @JimDeanSpivey `CheckstyleValidator can ensure that`.

Comment (pinaf) [inline qulice-checkstyle/src/test/java/com/qulice/checkstyle/CheckstyleValidatorTest.java]: @JimDeanSpivey typo `of of`

File: qulice-checkstyle/src/test/resources/com/qulice/checkstyle/AnonymousInnerClassPass.java [hunk 15]
```diff
@@ -0,0 +1,41 @@
...
+ */
+package foo;
+
+/**
+ * Simple.
+ * @author John Smith (john@example.com)
+ * @version $Id$
+ * @since 1.0
+ */
+public final class AnonymousInnerClassPass {
```
Comment (pinaf) [inline qulice-checkstyle/src/test/resources/com/qulice/checkstyle/AnonymousInnerClassPass.java]: @JimDeanSpivey `BlankLinesOutsideMethodsPass`

File: qulice-checkstyle/src/test/resources/com/qulice/checkstyle/AnonymousInnerClassFail.java [hunk 16]
```diff
@@ -0,0 +1,57 @@
...
+ */
+package foo;
+
+/**
+ * Simple.
+ * @author John Smith (john@example.com)
+ * @version $Id$
+ * @since 1.0
+ */
+public final class AnonymousInnerClassFail {
```
Comment (pinaf) [inline qulice-checkstyle/src/test/resources/com/qulice/checkstyle/AnonymousInnerClassFail.java]: @JimDeanSpivey `BlankLinesInsideMethodsFail`

File: qulice-checkstyle/src/main/java/com/qulice/checkstyle/EmptyLinesCheck.java [hunk 17]
```diff
@@ -53,29 +58,75 @@
...
+                final int last = closing.getLineNo();
+                this.methods.add(new LineRange(first, last));
             }
         }
     }
 
+    @Override
+    public void finishTree(final DetailAST root) {
+        final String[] lines = this.getLines();
+        for (int line = 0; line < lines.length; line += 1) {
```
Comment (pinaf) [inline qulice-checkstyle/src/main/java/com/qulice/checkstyle/EmptyLinesCheck.java]: @JimDeanSpivey `++line`

**Comment by pinaf** [comment]
@JimDeanSpivey please see 19 comments above.

File: qulice-checkstyle/src/main/java/com/qulice/checkstyle/EmptyLinesCheck.java (hunk 1 above)
Comment (JimDeanSpivey) [inline qulice-checkstyle/src/main/java/com/qulice/checkstyle/EmptyLinesCheck.java]: @pinaf It is thread safe, fixed description.

File: qulice-checkstyle/src/main/java/com/qulice/checkstyle/LineRanges.java (hunk 11 above)
Comment (JimDeanSpivey) [inline qulice-checkstyle/src/main/java/com/qulice/checkstyle/LineRanges.java]: @pinaf I would have to break encapsulation and expose the internal reference to the collection since `foreach` can only iterate over an `Iterable` and not an `Iterator`.

File: qulice-checkstyle/src/main/java/com/qulice/checkstyle/LineRanges.java (hunk 13 above)
Comment (JimDeanSpivey) [inline qulice-checkstyle/src/main/java/com/qulice/checkstyle/LineRanges.java]: @pinaf Without this nullcheck, findbugs complains:

```
[WARNING] FindBugs: M D NP: range must be non-null but is marked as nullable  At LineRanges.java:[line 129]
```

I think the issue here is that `Predicate` can iterate over collection's values and Java collections considers `null` an acceptable element.

**Comment by JimDeanSpivey** [comment]
@pinaf See updated code and comment replies.

File: qulice-checkstyle/src/main/java/com/qulice/checkstyle/LineRange.java [hunk 18]
```diff
@@ -0,0 +1,98 @@
...
+     * @param line The given line number to check.
+     * @return True if the given line number is within this range.
+     */
+    public boolean within(final int line) {
+        return line >= this.first() && line <= this.last();
+    }
+
+    /**
+     * Is the given range entirely within the LineRange. Example, given a
+     * LineRange of [10, 50], the proposed range of [12,48] should be within
```
Comment (pinaf) [inline qulice-checkstyle/src/main/java/com/qulice/checkstyle/LineRange.java]: @JimDeanSpivey `proposed` -> `given`

File: qulice-checkstyle/src/main/java/com/qulice/checkstyle/LineRanges.java [hunk 19]
```diff
@@ -0,0 +1,153 @@
...
+package com.qulice.checkstyle;
+
+import com.google.common.base.Predicate;
+import com.google.common.collect.FluentIterable;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Iterator;
+
+/**
+ * Represent a set of LineRange objects. For example, an instance of this class
```
Comment (pinaf) [inline qulice-checkstyle/src/main/java/com/qulice/checkstyle/LineRanges.java]: @JimDeanSpivey Represent`s`

File: qulice-checkstyle/src/main/java/com/qulice/checkstyle/LineRanges.java [hunk 20]
```diff
@@ -0,0 +1,153 @@
...
+        }
+
+        @Override
+        public boolean apply(final LineRange range) {
+            return range != null && range.within(this.given);
+        }
+    }
+
+    /**
+     * Thread safety via ThreadLocal.
```
Comment (pinaf) [inline qulice-checkstyle/src/main/java/com/qulice/checkstyle/LineRanges.java]: @JimDeanSpivey `Thread-safe collection of line ranges.`

**Comment by pinaf** [comment]
@JimDeanSpivey thanks! 3 more and we are good.

**Comment by JimDeanSpivey** [comment]
@pinaf The 3 updates were made, thanks.

**Comment by pinaf** [comment]
@rultor merge

File: qulice-checkstyle/src/main/java/com/qulice/checkstyle/EmptyLinesCheck.java [hunk 21]
```diff
@@ -53,29 +58,75 @@
...
+    /**
+     * If this is within a valid anonymous class, make sure that is still
+     * directly inside of a method of that anonymous inner class.
+     * Note: This implementation only checks one level deep, as nesting
+     * anonymous inner classes should never been done.
+     * @param line The line to check if it is within a method or not.
+     * @return True if the line is directly inside of a method.
+     */
+    private boolean validInnerClassMethod(final int line) {
+        return !this.anons.inRange(line)
```
Comment (krzyk) [inline qulice-checkstyle/src/main/java/com/qulice/checkstyle/EmptyLinesCheck.java]: @JimDeanSpivey this checks that the method is NOT within `anons` range?

File: qulice-checkstyle/src/main/java/com/qulice/checkstyle/LineRange.java [hunk 22]
```diff
@@ -0,0 +1,98 @@
...
+ *
+ * @author Jimmy Spivey (JimDeanSpivey@gmail.com)
+ * @version $Id$
+ */
+public final class LineRange {
+
+    /**
+     * The first (alpha) line number in the range.
+     */
+    private final transient int alpha;
```
Comment (krzyk) [inline qulice-checkstyle/src/main/java/com/qulice/checkstyle/LineRange.java]: @JimDeanSpivey why `alpha` and `omega` instead of `first` and `last`?

File: qulice-checkstyle/src/test/java/com/qulice/checkstyle/CheckstyleValidatorTest.java [hunk 23]
```diff
@@ -470,6 +470,49 @@ public void doesNotRejectUrlsInLongLines() throws Exception {
...
+    @Test
+    public void doesNotRejectSpaceInsideAnonClass() throws Exception {
+        this.validateCheckstyle(
+            "BlankLinesOutsideMethodsPass.java", true,
+            Matchers.containsString(CheckstyleValidatorTest.NO_VIOLATIONS)
+        );
+    }
+
+    /**
+     * Test if spaces are permitted between anonymous inner classes's methods.
```
Comment (krzyk) [inline qulice-checkstyle/src/test/java/com/qulice/checkstyle/CheckstyleValidatorTest.java]: @JimDeanSpivey *CheckstyleValidator can't allow spaces inside methods ....*

File: qulice-checkstyle/src/test/java/com/qulice/checkstyle/CheckstyleValidatorTest.java [hunk 24]
```diff
@@ -470,6 +470,49 @@ public void doesNotRejectUrlsInLongLines() throws Exception {
     }
 
     /**
+     * CheckstyleValidator can ensure that spaces can exist outside of methods
```
Comment (krzyk) [inline qulice-checkstyle/src/test/java/com/qulice/checkstyle/CheckstyleValidatorTest.java]: @JimDeanSpivey *CheckstyleValidator can allow spaces between methods of anonymous classes.*
(http://www.yegor256.com/2014/04/27/typical-mistakes-in-java-code.html#test-method-names)

File: qulice-checkstyle/src/main/java/com/qulice/checkstyle/EmptyLinesCheck.java (hunk 21 above)
Comment (JimDeanSpivey) [inline qulice-checkstyle/src/main/java/com/qulice/checkstyle/EmptyLinesCheck.java]: The full method body wasn't shown:

```java
    private boolean validInnerClassMethod(final int line) {
        return !this.anons.inRange(line)
            || this.methods.within(this.anons).inRange(line);
    }
```

It is checking 2 criteria:
1. If the given line is not within an anonymous inner type, then bypass this validation
2. Otherwise, is it within an **anonymous inner type's method** ? If so, return true so that the empty line check regex test can be performed.

File: qulice-checkstyle/src/test/java/com/qulice/checkstyle/CheckstyleValidatorTest.java [hunk 25]
```diff
@@ -470,6 +470,49 @@ public void doesNotRejectUrlsInLongLines() throws Exception {
...
+     * that are within anonymous inner classes.
+     * @throws Exception In case of error
+     */
+    @Test
+    public void doesNotRejectSpaceInsideAnonClass() throws Exception {
```
Comment (krzyk) [inline qulice-checkstyle/src/test/java/com/qulice/checkstyle/CheckstyleValidatorTest.java]: @JimDeanSpivey `allowsSpacesBetwenMethodsOfAnonymousClasses`

File: qulice-checkstyle/src/test/java/com/qulice/checkstyle/CheckstyleValidatorTest.java [hunk 26]
```diff
@@ -470,6 +470,49 @@ public void doesNotRejectUrlsInLongLines() throws Exception {
...
+            Matchers.containsString(CheckstyleValidatorTest.NO_VIOLATIONS)
+        );
+    }
+
+    /**
+     * Test if spaces are permitted between anonymous inner classes's methods.
+     * @throws Exception In case of error
+     */
+    @Test
+    public void doesRejectSpaceBetweenAllMethods() throws Exception {
```
Comment (krzyk) [inline qulice-checkstyle/src/test/java/com/qulice/checkstyle/CheckstyleValidatorTest.java]: @JimDeanSpivey `rejectsSpacesInsideMethods`

File: qulice-checkstyle/src/main/java/com/qulice/checkstyle/LineRange.java [hunk 27]
```diff
@@ -0,0 +1,98 @@
...
+    public boolean within(final LineRange range) {
+        return range.first() >= this.first()
+            && range.last()  <= this.last();
+    }
+
+    /**
+     * Get the alpha line number.
+     * @return The alpha line number.
+     */
+    public int first() {
```
Comment (krzyk) [inline qulice-checkstyle/src/main/java/com/qulice/checkstyle/LineRange.java]: @JimDeanSpivey you don't need this method

File: qulice-checkstyle/src/main/java/com/qulice/checkstyle/LineRange.java (hunk 22 above)
Comment (JimDeanSpivey) [inline qulice-checkstyle/src/main/java/com/qulice/checkstyle/LineRange.java]: I'm forced to create doppleganger variable names because the constructor won't allow this:

```java
    public LineRange(final int first, final int last) {
        this.first = first;
        this.first = last;
    }
```

Also, due to qulice conventions, I am not able to have these getter names, as instance variables cannot have the same name as method names:

```java
    public int first() {
        return this.alpha;
    }

    public int last() {
        return this.omega;
    }
```

I am under the impression that it is better to name the parameters to the constructor first/last, and the getters named first/last and let the encapsulated fields have the more unusual name.

File: qulice-checkstyle/src/main/java/com/qulice/checkstyle/LineRange.java [hunk 28]
```diff
@@ -0,0 +1,98 @@
...
+        return this.alpha;
+    }
+
+    /**
+     * Get the omega line number.
+     * @return The omega line number.
+     */
+    public int last() {
```
Comment (krzyk) [inline qulice-checkstyle/src/main/java/com/qulice/checkstyle/LineRange.java]: @JimDeanSpivey and this is also not needed

**Comment by krzyk** [comment]
@JimDeanSpivey please see comments from me

File: qulice-checkstyle/src/main/java/com/qulice/checkstyle/EmptyLinesCheck.java (hunk 21 above)
Comment (krzyk) [inline qulice-checkstyle/src/main/java/com/qulice/checkstyle/EmptyLinesCheck.java]: @JimDeanSpivey OK, thanks

File: qulice-checkstyle/src/main/java/com/qulice/checkstyle/LineRange.java (hunk 22 above)
Comment (krzyk) [inline qulice-checkstyle/src/main/java/com/qulice/checkstyle/LineRange.java]: @JimDeanSpivey constructor should allow this (I'm not sure which qulice version has this updated to allow constructor params same as fields), but having method names same as fields is not allowed so yeah, it shoudl stay as it is.

**Comment by JimDeanSpivey** [comment]
@krzyk I believe everything has been updated per your comments, let me know if I missed something.

**Comment by krzyk** [comment]
@JimDeanSpivey Thanks, nice job.

**Comment by krzyk** [comment]
@rultor merge pls

**Comment by krzyk** [comment]
@rultor merge pls

**Comment by davvd** [comment]
@pinaf paid, thanks, added **40 mins** to your account, payment `AP-6XA02595X7274341K`, 70 hours and 43 mins was spent. extra minutes for review comments (c=25). +40 to your rating, your total score is [+6740](http://www.netbout.com/b/35128?open=rating)

**Comment by davvd** [comment]
@rultor please deploy
